iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/Tools.java

```diff
@@ -137,99 +154,440 @@ public static RegistryManager getRegistyManager(String iotHubConnectionString) t
         return registryManager;
     }
 
-    public static TestDeviceIdentity getTestDevice(String iotHubConnectionString, IotHubClientProtocol protocol, AuthenticationType authenticationType) throws URISyntaxException, IOException, IotHubException, GeneralSecurityException, InterruptedException
+    private static Queue<TestDeviceIdentity> testSasDeviceQueue = new ConcurrentLinkedQueue<>();  // queue contains SAS device identities that are ready to be re-used, and that have no twin modifications yet
+    private static Queue<TestDeviceIdentity> testX509DeviceQueue = new ConcurrentLinkedQueue<>(); // queue contains x509 device identities that are ready to be re-used, and that have no twin modifications yet
+    private static Queue<TestDeviceIdentity> testSasDeviceWithTwinQueue = new ConcurrentLinkedQueue<>(); // queue contains SAS device identities that are ready to be re-used, and that have some twin modifications to them already
```

Comment:
If the number of tests that need a fresh twin is very low, I'm wondering if we really need 3 queues. We could probably just get a new identity instead of adding more logic to this class.

Reply:
Unfortunately the number of tests that need a fresh twin is rather high. I'd estimate that we have ~400 test runs that assume no initial twin is present. We have 

number of twin operations (desired property tests, query, reported property, getTwin, fault injection for each as well)

times

number of twin protocols (4)

times

device client vs module client (2)

Reply:
The real benefit here is to allow an identity used in a twin test to be re-used in a telemetry/methods/c2d message test